_posts/2019-05-21-i18n-messages.adoc

```diff
@@ -0,0 +1,327 @@
+
+=== Referencing Messages in XML Descriptors
+
+The first option is to reference a particular message key within XML UI descriptor. In the Petclinic example, the caption of the "regular checkup" button should be displayed localized. This happens through the prefix `msg://` followed by the translation key:
+
+.visit-browse.xml
+[source,xml]
+----
+<groupTable id="visitsTable"
+            dataContainer="visitsDc"
+            width="100%">
+    <actions>
+
+        <action id="createRegularCheckup"
+                caption="msg://createRegularCheckup" // <1>
+                icon="font-icon:BRIEFCASE" />
+
+    </actions>
+    <buttonsPanel id="buttonsPanel"
+                  alwaysVisible="true">
+        <button id="createForPetBtn" action="visitsTable.createRegularCheckup"/>
+    </buttonsPanel>
+</groupTable>
+----
+<1> `msg://createRegularCheckup` references the key `createRegularCheckup` in the current package of the XML descriptor
+
+=== Programmatic Localization of Messages
+
+The second way of interacting with the Localization mechanism is imperatively through Java code. In the Petclinic example, the "regular checkup" functionality should render the following two dynamic messages programmatically:
+
+1. when a new Regular Checkup screen is opened, the description attribute for the visit entity should be pre-filled with a particular localized message
+2. once the Regular Checkup is created, a localized notification message should be rendered, containing the name of the pet
+
+Messages can be received programmatically via the `Messages` Interface. It provides an API to fetch a translated String through the localization mechanism of CUBA for the locale of the current user.
+
+TIP: In the example code instead of using the `Messages` Interface, the `MessageBundle` Interface is used. It is a new shorthand mechanism introduced in CUBA 7.0, that fetches a particular message from the message pack of the calling Java class. This way it is not required to provide the package for every translation API request.
+
+
+The first localized message for pre-filling the description attribute is created as follows:
+
+
+.VisitBrowse.java
+[source,java]
+----
+public class VisitBrowse extends StandardLookup<Visit> {
+
+    @Inject
+    private Notifications notifications;
+
+    @Inject
+    private MessageBundle messageBundle;
+
+    // ...
+
+    private void createVisitForPet(Pet pet) {
+        screenBuilders.editor(visitsTable)
+                .newEntity()
+                .withInitializer(visit -> {
+                    visit.setDescription(
+                        regularCheckupDescriptionContent(pet) // <1>
+                    );
+                    visit.setPet(pet);
+                })
+                .withScreenClass(RegularCheckup.class)
+                .withLaunchMode(OpenMode.DIALOG)
+                .withAfterCloseListener(event -> showRegularCheckupCreatedNotification(pet))
```

Comment:
This line does not fit in the resulting page, please make sure here and in other parts the lines are not wrapped.

Reply:
👍